# 03 – ML Models
**Owner:** TV2

**Mục tiêu:** chạy 5 model ML: 
- MODEL_01: Linear Regression
- MODEL_02: Decision Tree Regressor
- MODEL_03: Random Forest Regressor
- MODEL_04: Gradient Boosting Regressor
- MODEL_05: XGBoost Regressor


**Output:** metrics và experiment records.


In [8]:
"""
import sys
sys.path.append("../src")
from preprocessing import load_processed_data
from ml_models import train_one_model, predict_price
from evaluation import regression_metrics

X_train, X_val, y_train, y_val, X_test = load_processed_data("../data/processed")
model_ids = ["MODEL_01","MODEL_02","MODEL_03","MODEL_04","MODEL_05"]
results = []
for model_id in model_ids:
    model, training_time = train_one_model(model_id, X_train, y_train)
    val_pred = predict_price(model, X_val)
    metrics = regression_metrics(y_val, val_pred)
    print(model_id, metrics, training_time)
    results.append({"model": model_id, **metrics,
                    "training_time_seconds": training_time,
                    "kaggle_score": None, "notes": ""})
"""


'\nimport sys\nsys.path.append("../src")\nfrom preprocessing import load_processed_data\nfrom ml_models import train_one_model, predict_price\nfrom evaluation import regression_metrics\n\nX_train, X_val, y_train, y_val, X_test = load_processed_data("../data/processed")\nmodel_ids = ["MODEL_01","MODEL_02","MODEL_03","MODEL_04","MODEL_05"]\nresults = []\nfor model_id in model_ids:\n    model, training_time = train_one_model(model_id, X_train, y_train)\n    val_pred = predict_price(model, X_val)\n    metrics = regression_metrics(y_val, val_pred)\n    print(model_id, metrics, training_time)\n    results.append({"model": model_id, **metrics,\n                    "training_time_seconds": training_time,\n                    "kaggle_score": None, "notes": ""})\n'

## TODO TV2
- Hoàn thiện 5 model.
- Điều chỉnh một số hyperparameter nếu cần.
- Ghi experiment.
- Không thêm model thứ 6.


# Machine Learning Models (5 Models) — House Prices

## 1. Checklist
- Load dữ liệu processed từ TV1
- Kiểm tra shape và kiểu dữ liệu
- Khởi tạo 5 mô hình
- Training
- Tính RMSLE, RMSE, MAE, R2.
- Đo training time.
- Thử một vài hyperparameter hợp lý 
- Kết quả sau sau thử thay đổi tham số
- Bàn giao kết quả cho TV4

## 2. Import thư viện

In [9]:
# Cho phép người dùng import trực tiếp các file .py hoặc package nằm trong thư mục
import sys
sys.path.append("D:\\House_Price")

import numpy as np
import time
import ml_models
import importlib # Load lại module ml_models để áp dụng các thay đổi mới nhất
importlib.reload(ml_models)


from preprocessing import load_processed_data
from ml_models import train_one_model, predict_price
from evaluation import regression_metrics


## 3. Load dữ liệu

Load và xem các tệp chứa bao nhiêu mẫu

In [10]:
X_train, X_val, y_train, y_val, X_test = load_processed_data("D:\\Zalo Received Files\\Deep learning\\Tuần 02\\House_Price\\data\\processed")

#Kiểm tra dữ liệu đầu vào
print("Kiểm tra dữ liệu đầu vào:")
print("X_train:", X_train.shape)
print("X_val  :", X_val.shape)
print("y_train:", y_train.shape)
print("y_val  :", y_val.shape)
print("X_test :", X_test.shape)

Kiểm tra dữ liệu đầu vào:
X_train: (1168, 286)
X_val  : (292, 286)
y_train: (1168,)
y_val  : (292,)
X_test : (1459, 286)


## 4. Kiểm tra dữ liệu

In [11]:
print("X_train dtype:", X_train.dtype)
print("X_val dtype  :", X_val.dtype)
print("y_train dtype:", y_train.dtype)
print("y_val dtype  :", y_val.dtype)

print("\nMissing values:")
print("X_train NaN:", np.isnan(X_train).sum())
print("X_val NaN  :", np.isnan(X_val).sum())
print("y_train NaN:", np.isnan(y_train).sum())
print("y_val NaN  :", np.isnan(y_val).sum())

print("\nFeature dimension:", X_train.shape[1])

X_train dtype: float64
X_val dtype  : float64
y_train dtype: int64
y_val dtype  : int64

Missing values:
X_train NaN: 0
X_val NaN  : 0
y_train NaN: 0
y_val NaN  : 0

Feature dimension: 286


## 5. Training 5 models

In [12]:
from ml_models import build_model, train_one_model, predict_price


# Xác định các ID mô hình
model_ids = ["MODEL_01","MODEL_02","MODEL_03","MODEL_04","MODEL_05"]

# Hàm lấy parameters của model
def get_model_params(model_id):
    model = build_model(model_id)

    params = model.get_params()

    # Giữ parameter có giá trị khác None
    params = {
        key: value
        for key, value in params.items()
        if value is not None
    }

    return params

# Training và evaluation
results = []

for model_id in model_ids:
    # Lấy parameters
    params = get_model_params(model_id)

    # Train model
    model, training_time_seconds = train_one_model(
        model_id,
        X_train,
        y_train
    )

    # Predict validation set
    val_pred = predict_price(
        model,
        X_val
    )

    # Tính metrics
    metrics = regression_metrics(
        y_val,
        val_pred
    )

    # Print kết quả
    print("=" * 80)
    print("MODEL:", model_id)
    print("Parameters:", params)
    print("Metrics:", metrics)
    print("Training time:", training_time_seconds)
    print("=" * 80)

    # Save kết quả
    results.append({
        "model": model_id,
        "parameters": params,
        "metrics": metrics,
        "training_time_seconds": training_time_seconds,
        "kaggle_score": None,
        "notes": ""
    })

MODEL: MODEL_01
Parameters: {'copy_X': True, 'fit_intercept': True, 'positive': False, 'tol': 1e-06}
Metrics: {'RMSLE': 0.18736358741920328, 'RMSE': 29477.004825239714, 'MAE': 18284.896325872887, 'R2': 0.886720096382205}
Training time: 0.07669319999877189
MODEL: MODEL_02
Parameters: {'ccp_alpha': 0.0, 'criterion': 'squared_error', 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 42, 'splitter': 'best'}
Metrics: {'RMSLE': 0.2185261668478426, 'RMSE': 41853.012719622755, 'MAE': 27668.29109589041, 'R2': 0.7716297030408461}
Training time: 0.07396780000090075
MODEL: MODEL_03
Parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'criterion': 'squared_error', 'max_features': 1.0, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 300, 'n_jobs': -1, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
Metrics: {'RMSLE': 0.15235

## 6. Thử các mô hình bằng tham số khác


In [ ]:
import random
#Load lại module ml_models để áp dụng các thay đổi mới nhất
importlib.reload(ml_models)

# Tạo các hyperparameters ngẫu nhiên cho các mô hình khác nhau
def random_hyperparameters(model_id):

    if model_id == "MODEL_02":

        return {
            "max_depth": random.choice([None, 5, 10, 15, 20, 25, 30]),
            "min_samples_split": random.choice([2, 5, 10, 15, 20]),
            "min_samples_leaf": random.choice([1, 2, 4, 8, 10])
        }


    elif model_id == "MODEL_03":

        return {
            "n_estimators": random.choice([100, 200, 300, 400, 500]),
            "max_depth": random.choice([None, 5, 10, 15, 20]),
            "min_samples_split": random.choice([2, 5, 10, 15]),
            "min_samples_leaf": random.choice([1, 2, 4, 8])
        }


    elif model_id == "MODEL_04":

        return {
            "n_estimators": random.choice([100, 200, 300, 400, 500]),
            "learning_rate": random.choice([
                0.01, 0.03, 0.05, 0.1
            ]),
            "max_depth": random.choice([2, 3, 4, 5]),
            "min_samples_split": random.choice([2, 5, 10, 15]),
            "min_samples_leaf": random.choice([1, 2, 4, 8])
        }


    elif model_id == "MODEL_05":

        return {
            "n_estimators": random.choice([100, 200, 300, 500, 700]),
            "learning_rate": random.choice([
                0.01, 0.03, 0.05, 0.1
            ]),
            "max_depth": random.choice([3, 4, 5, 6, 8]),
            "min_child_weight": random.choice([1, 3, 5]),
            "subsample": random.choice([
                0.7, 0.8, 0.9, 1.0
            ]),
            "colsample_bytree": random.choice([
                0.7, 0.8, 0.9, 1.0
            ])
        }

    else:

        raise ValueError(
            f"Random hyperparameters are not defined "
            f"for {model_id}"
        )

In [17]:
model_ids = ["MODEL_01", "MODEL_02", "MODEL_03", "MODEL_04", "MODEL_05"]

results = []

for model_id in model_ids:

    # Lấy object model ra mà không sửa hàm gốc
    model, training_time_seconds = train_one_model(
        model_id,
        X_train,
        y_train
    )

    # Lấy params ngẫu nhiên
    try:
        params = random_hyperparameters(model_id)
    except ValueError:
        params = {}

    # Gán đè params ngẫu nhiên vào model và fit lại để tính lại thời gian chuẩn
    if params:
        model.set_params(**params)
        
        start_time = time.time()
        model.fit(X_train, y_train)
        training_time_seconds = time.time() - start_time

    # In hyperparameters
    print("=" * 80)
    print(f"MODEL: {model_id}")
    print("Parameters:", {
        key: value
        for key, value in model.get_params().items()
        if key in [
            "n_estimators",
            "max_depth",
            "learning_rate",
            "min_samples_split",
            "min_samples_leaf",
            "min_child_weight",
            "subsample",
            "colsample_bytree",
            "random_state",
            "n_jobs"
        ]
    })

    # Predict validation
    val_pred = predict_price(
        model,
        X_val
    )

    # Evaluation
    metrics = regression_metrics(
        y_val,
        val_pred
    )

    # Print results
    print()
    print("Metrics:")
    print(f"  RMSLE: {metrics['RMSLE']:.5f}")
    print(f"  RMSE:  {metrics['RMSE']:.2f}")
    print(f"  MAE:   {metrics['MAE']:.2f}")
    print(f"  R2:    {metrics['R2']:.5f}")
    print(f"  Training Time: {training_time_seconds:.4f} seconds")
    print("=" * 80)
    print()

    # Lưu kết quả 
    results.append({
        "model": model_id,
        "metrics": metrics,
        "training_time_seconds": training_time_seconds,
        "parameters": params,
        "kaggle_score": None,
        "notes": ""
    })

MODEL: MODEL_01
Parameters: {'n_jobs': None}

Metrics:
  RMSLE: 0.18736
  RMSE:  29477.00
  MAE:   18284.90
  R2:    0.88672
  Training Time: 0.0760 seconds

MODEL: MODEL_02
Parameters: {'max_depth': 25, 'min_samples_leaf': 8, 'min_samples_split': 20, 'random_state': 42}

Metrics:
  RMSLE: 0.18966
  RMSE:  38812.68
  MAE:   24154.41
  R2:    0.80360
  Training Time: 0.0320 seconds

MODEL: MODEL_03
Parameters: {'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 100, 'n_jobs': -1, 'random_state': 42}

Metrics:
  RMSLE: 0.15415
  RMSE:  29112.21
  MAE:   17552.97
  R2:    0.88951
  Training Time: 0.8579 seconds

MODEL: MODEL_04
Parameters: {'learning_rate': 0.05, 'max_depth': 5, 'min_samples_leaf': 4, 'min_samples_split': 15, 'n_estimators': 400, 'random_state': 42, 'subsample': 1.0}

Metrics:
  RMSLE: 0.14075
  RMSE:  25640.09
  MAE:   16083.25
  R2:    0.91429
  Training Time: 9.3470 seconds

MODEL: MODEL_05
Parameters: {'colsample_bytree': 0.7, 'learning_ra

### 7. Kết luận

- **MODEL_02 – Decision Tree:** Sau khi điều chỉnh tham số, mô hình có sự cải thiện. RMSLE, RMSE và MAE giảm nhẹ. Trong khi R² tăng từ 0.771630 lên 0.803600.
- **MODEL_03 – Random Forest:** Sau khi thay đổi tham số, các chỉ số dự đoán đều giảm nhẹ về chất lượng: RMSLE, RMSE và MAE tăng, R² giảm. Tuy nhiên, thời gian huấn luyện giảm đáng kể từ 3.74 xuống 1.63 giây.
- **MODEL_04 – Gradient Boosting:** Mô hình gần như không thay đổi tuy nhiên sau khi điều chỉnh tham số thì thời gian huấn luyện tăng đáng kể.
- **MODEL_05 – XGBoost:** Mô hình cũng không có sự thay đổi quá nhiều tuy nhiên thời gian huấn luyện lại giảm đáng kể.